crates/autopilot/src/infra/persistence/mod.rs

```diff
@@ -155,12 +159,164 @@ impl Persistence {
+                    .map(|o| (auction_id, ByteArray(o.0)))
+                    .collect::<Vec<_>>()
+                    .as_slice(),
+            )
+            .await
+            .context("fetch fee policies")
+            .map_err(error::Auction::DbError)?
+            .into_iter()
+            .map(|((_, order), policies)| (domain::OrderUid(order.0), policies))
+            .collect::<HashMap<_, _>>();
+
+            // get quotes for all orders with PriceImprovement fee policy
+            let quotes = self
+                .postgres
+                .read_quotes(fee_policies.iter().filter_map(|(order_uid, policies)| {
+                    policies
+                        .iter()
+                        .any(|policy| {
+                            matches!(
+                                policy.kind,
+                                database::fee_policies::FeePolicyKind::PriceImprovement
+                            )
+                        })
+                        .then_some(order_uid)
+                }))
+                .await
+                .map_err(error::Auction::DbError)?;
+
+            // compile order data
+            let mut orders = HashMap::new();
+            for order in auction_orders.iter() {
+                let order_policies = match fee_policies.get(order) {
+                    Some(policies) => policies
+                        .iter()
+                        .cloned()
+                        .map(|policy| {
+                            dto::fee_policy::try_into_domain(policy, quotes.get(order))
+                                .context(format!("convert fee policy for order {}", order))
+                                .map_err(error::Auction::FeePolicy)
+                        })
+                        .collect::<Result<Vec<_>, _>>()?,
+                    None => vec![],
+                };
+                orders.insert(*order, order_policies);
+            }
+            orders
+        };
+
+        Ok(domain::settlement::Auction {
+            id: auction_id,
+            orders,
+            prices,
+            deadline,
+            surplus_capturing_jit_order_owners,
+        })
+    }
+}
+
+#[derive(prometheus_metric_storage::MetricStorage)]
+struct Metrics {
+    /// Timing of db queries.
+    #[metric(name = "persistence_database_queries", labels("type"))]
+    database_queries: prometheus::HistogramVec,
+}
+
+impl Metrics {
+    fn get() -> &'static Self {
+        Metrics::instance(observe::metrics::get_storage_registry()).unwrap()
+    }
 }
 
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
     #[error("failed to read data from database")]
     DbError(#[from] anyhow::Error),
-    #[error(transparent)]
-    Price(#[from] domain::auction::InvalidPrice),
+}
+
+pub mod error {
+    use super::*;
+
+    #[derive(Debug, thiserror::Error)]
+    pub enum Auction {
+        #[error("failed to read data from database: {0}")]
+        DbError(#[source] anyhow::Error),
```

Comment:
What does `#[source]` do and why is it useful here?

Reply:
@fleupold `#[source]`  is used primarily for error chaining, helping you to trace back through nested errors. If we were to check an upper level which was the original cause, we couldn't see it, as `#[from}` squashes the original error. You can actually chain both.

@sunce86 implemented `#[source]` because of my suggestion. I personally always like to implement `#[source]` not to lose the original error and have all nested properly.

Reply:
This is extremely helpful/powerful if we use `thiserror` crate all the time (since it forces you to define an enum per error instead of just a string, making it more robust and easy to act on specific errors)